### ISSUE TITLE
Automatic blocksize for dask.dataframe.read_csv

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1147]
We could automatically tune the default blocksize based on total physical or available memory and the number of cores.  This would avoid memory errors for smaller machines and novice users.

This would probably depend on the optional presence of `psutil`

```
psutil.virtual_memory()
psutil.cpu_count()
```

**Comment by ijstokes** [comment]
My scenario: 5.7M rows of CSV data from this file: http://stat-computing.org/dataexpo/2009/2000.csv.bz2 For context, see stat-computing.org/dataexpo/2009/the-data.html

I have a 2 GB RAM VM and I can't get the following code to run without crashing a Jupyter Kernel:

```
df = dd.read_csv('data/2000.csv', header=0, names=cols, dtype=dtypes, blocksize=int(1e7))
expr = df.depdelay.groupby(df.origin).mean().nlargest(n=10)
with Profiler() as prof, ResourceProfiler(dt=0.1) as rprof:
    res = expr.compute()
```

I've tried a wide range of blocksizes from 1e5 to 1e8

**Comment by ijstokes** [comment]
Here is the first time I've done better than a dead kernel.  Something failed and I got a stack-trace instead, in case this is helpful:

```
[                                        ] | 0% Completed |  1.4s

---------------------------------------------------------------------------
CParserError                              Traceback (most recent call last)
<ipython-input-6-31ed29cfd077> in <module>()
      1 #with Profiler() as prof, ResourceProfiler(dt=0.1) as rprof:
----> 2 res = expr.compute()

/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/base.pyc in compute(self, **kwargs)
     35 
     36     def compute(self, **kwargs):
---> 37         return compute(self, **kwargs)[0]
     38 
     39     @classmethod

/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/base.pyc in compute(*args, **kwargs)
    108                 for opt, val in groups.items()])
    109     keys = [var._keys() for var in variables]
--> 110     results = get(dsk, keys, **kwargs)
    111 
    112     results_iter = iter(results)

/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/threaded.pyc in get(dsk, result, cache, num_workers, **kwargs)
     55     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     56                         cache=cache, queue=queue, get_id=_thread_get_id,
---> 57                         **kwargs)
     58 
     59     return results

/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/async.pyc in get_async(apply_async, num_workers, dsk, result, cache, queue, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, **kwargs)
    486                 _execute_task(task, data)  # Re-execute locally
    487             else:
--> 488                 raise(remote_exception(res, tb))
    489         state['cache'][key] = res
    490         finish_task(dsk, key, state, results, keyorder.get)

CParserError: get_plugin

Traceback
---------
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/async.py", line 267, in execute_task
    result = _execute_task(task, data)
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/async.py", line 249, in _execute_task
    return func(*args2)
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/dask/dataframe/csv.py", line 41, in bytes_read_csv
    df = pd.read_csv(bio, **kwargs)
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/pandas/io/parsers.py", line 498, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/pandas/io/parsers.py", line 285, in _read
    return parser.read()
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/pandas/io/parsers.py", line 747, in read
    ret = self._engine.read(nrows)
  File "/Users/ijstokes/anaconda/envs/trialbox/lib/python2.7/site-packages/pandas/io/parsers.py", line 1197, in read
    data = self._reader.read(nrows)
  File "pandas/parser.pyx", line 766, in pandas.parser.TextReader.read (pandas/parser.c:7988)
  File "pandas/parser.pyx", line 788, in pandas.parser.TextReader._read_low_memory (pandas/parser.c:8244)
  File "pandas/parser.pyx", line 842, in pandas.parser.TextReader._read_rows (pandas/parser.c:8970)
  File "pandas/parser.pyx", line 829, in pandas.parser.TextReader._tokenize_rows (pandas/parser.c:8838)
  File "pandas/parser.pyx", line 1833, in pandas.parser.raise_parser_error (pandas/parser.c:22649)
```

**Comment by doctorf72** [comment]
Seems to be one of the solutions:
Function returns in the runtime avaliable RAM volume:

import psutil
def GetUserMemory():
    Memory = psutil.swap_memory()
    return Memory[2]

Please let me know if it works?

**Comment by mrocklin** [comment]
I recommend `psutil.virtual_memory().total`

**Comment by doctorf72** [comment]
In this case block size will be fixed this not guarantees proper allocation.this is my opinion.

**Comment by doctorf72** [comment]
import psutil
def GetUserMemory():
    Memory = psutil.virtual_memory().total
    return Memory

This function returning long type of available memory

**Comment by dukebody** [comment]
I understand that default blocksize should be `psutil.virtual_memory().total / psutil.cpu_count()`?

**Comment by mrocklin** [comment]
That, probably divided by a further factor of 10 or so.  Also, I can't imagine a case where we would want a blocksize greater than 64MB.

We want to avoid filling ram.  It's ok to be way less than total memory (factor of ten is great) and it's terrible to be greater than total memory (things run 100x slower).  A block of data on disk might very easily expand to 5x its original size when converted from CSV to a dataframe.  We will have about as many blocks in ram as we have cores.

**Comment by dukebody** [comment]
There is another implementation of a solution to this issue is this PR: https://github.com/dask/dask/pull/1215

**Comment by dukebody** [comment]
This is fixed by #1328 , let's close. :)
